增加 removeByDomainWithKeys 支援部份名稱或 * 前導名稱，並可指定多個參數必須同時存在才清除這些參數；版本增加到 1.13

ChrisTorng opened the pull request:
removeByDomain 增加支援部份名稱或 * 前導名稱 功能是由 ChatGPT 4o 更新。
很少有機會發 PR，希望沒有破壞或冒犯到什麼。

ChrisTorng committed:
TrackingTokenStripper.user.js 之 removeByDomain 增加支援部份名稱或 * 前導名稱；增加支援 *.substack.com 及 www.latent.space 網站去除追蹤碼；版本增加到 1.13。

doggy8088 commented:
以下這幾條規則可否提供一些範例網址？

```
// Substack, Latent Space
.removeByDomain('*.substack.com', 'publication_id')
.removeByDomain('*.substack.com', 'post_id')
.removeByDomain('*.substack.com', 'isFreemail')
.removeByDomain('*.substack.com', 'r')
.removeByDomain('*.substack.com', 'triedRedirect')

.removeByDomain('www.latent.space', 'publication_id')
.removeByDomain('www.latent.space', 'post_id')
.removeByDomain('www.latent.space', 'isFreemail')
.removeByDomain('www.latent.space', 'r')
.removeByDomain('www.latent.space', 'triedRedirect')
```

可否把「通配符」都改為「萬用字元」？

ChrisTorng commented:
謝謝您的提問，我找出原始的 Email 通知網址，發現還有優化空間，改善後再發 PR，不過要晚些。註解也會改善。

ChrisTorng committed:
TrackingTokenStripper.user.js 之 removeByDomain 擴充單一參數之 Name 為可接納多項目之 keys 參數。
Substack 相關網址設定統一一條移除。

ChrisTorng commented:
TrackingTokenStripper.user.js 之 removeByDomain 擴充單一參數之 Name 為可接納多項目之 keys 參數。
Substack 相關網址設定統一一條移除規則。
改善註解文字。

我有 Email 訂閱 [Substack](https://substack.com) 中一些作者，也包括自己的網站。
由於可自訂網域，故以 null domain 並增加同時符合多參數的比對模式。
底下提供一些 Email 中取得的網址:
https://substack.com/app-link/post?publication_id=1084089&post_id=153133301&utm_source=post-email-title&utm_campaign=email-post-title&isFreemail=true&r=1dcd1c&token=eyJ1c2VyX2lkIjo4Mjg3Nzk1MiwicG9zdF9pZCI6MTUzMTMzMzAxLCJpYXQiOjE3MzQyMTYyNTEsImV4cCI6MTczNjgwODI1MSwiaXNzIjoicHViLTEwODQwODkiLCJzdWIiOiJwb3N0LXJlYWN0aW9uIn0.IzhtF8VJCwlYJ8LzpmZmW-77HWyVEkA7hgD6YGmR9p0

https://substack.com/app-link/post?publication_id=1173386&post_id=149590578&utm_source=post-email-title&utm_campaign=email-post-title&isFreemail=true&r=1dcd1c&token=eyJ1c2VyX2lkIjo4Mjg3Nzk1MiwicG9zdF9pZCI6MTQ5NTkwNTc4LCJpYXQiOjE3Mjc2NjUwMzMsImV4cCI6MTczMDI1NzAzMywiaXNzIjoicHViLTExNzMzODYiLCJzdWIiOiJwb3N0LXJlYWN0aW9uIn0.-IkDSnrcHSW0UxZuZW2xAp-X519Eh0SfZTBmQNAejSg

https://substack.com/app-link/post?publication_id=1705980&post_id=149587042&utm_source=post-email-title&utm_campaign=email-post-title&isFreemail=false&r=1dcd1c&token=eyJ1c2VyX2lkIjo4Mjg3Nzk1MiwicG9zdF9pZCI6MTQ5NTg3MDQyLCJpYXQiOjE3Mjc2NTk4NTksImV4cCI6MTczMDI1MTg1OSwiaXNzIjoicHViLTE3MDU5ODAiLCJzdWIiOiJwb3N0LXJlYWN0aW9uIn0.sdnGADWZJ0Fipq1oOMRXEgjbAFoLd7uU5Jw0MCTjH84

https://email.mg-d1.substack.com/c/eJxkkM1uhSAQRp9GdhoYf4AFi6bJfYLuzQijkl7FwFjTt2-uqzZdn5ycL59HpiXlb3ekwnVh5FLThvEpgmvtJI0fBDmlwcoBDBhxw3GhnTIyhRH5F-1tJ1YHKoCiXgdtPSpjANu59cFIaZT33SSiAwmdkmqQWrXt0LTNrHG21qMJMwJpbGzp6cp5qDq5LXVQTTmnwug_G582Ecs4Z7q3OM4niadbmY9StW8VPCp4-DXHwinvyx-xgsdxU-Tl4FKfe6BcGPcQ96VOc72RyO79JX-85KqTa-JX5s6Wcwppw7i767oE_z_tLJTHGJwBo7XtQXw5-AkAAP__NNR0hQ

https://substack.com/app-link/post?publication_id=1705980&post_id=150246159&utm_source=substack&utm_medium=email&isFreemail=false&comments=true&token=eyJ1c2VyX2lkIjo4Mjg3Nzk1MiwicG9zdF9pZCI6MTUwMjQ2MTU5LCJpYXQiOjE3Mjg5NzI3NzksImV4cCI6MTczMTU2NDc3OSwiaXNzIjoicHViLTE3MDU5ODAiLCJzdWIiOiJwb3N0LXJlYWN0aW9uIn0.RBahZRkA5UiD1ofLi4xoacajlcVzuI6feYeUUVgmWmk&r=1dcd1c&utm_campaign=email-half-magic-comments&action=post-comment&utm_source=substack&utm_medium=email

因可同一網站比對多參數，因此先前既有的單一網站多條比對的規則，有可能各別網站合併為一條。不過自己沒有檢驗過，也不清楚如何取得它們的相關網址，因此暫時沒有改動它們。

剛剛仔細再看原規則，發現漏了支援規則:
```js
.removeByDomain('www.facebook.com', 'ref=watch_permalink')
```
另這本這條不確定怎麼運作的:
```js
.removeByDomain('www.facebook.com', 'ref', 'notif')
```
我還會再改善以支援原來的規則。

ChrisTorng committed:
為支援原本之 key=value 格式，故將修改的函式改名為 removeByDomainWithKeys 新函式，原 removeByDomain 函式加回。

ChrisTorng commented:
為支援原本之 key=value 格式，故將修改的函式改名為 removeByDomainWithKeys 新函式，原 removeByDomain 函式加回。

doggy8088 commented:
為什麼這行你不加上域名呢？直接移除 `r` 或 `token` 或 `post_id` 這樣常見的參數並套用到「所有網站」我覺得不可行。

```js
.removeByDomainWithKeys(null, 'publication_id', 'post_id', 'isFreemail', 'r', 'token', 'triedRedirect')
```

ChrisTorng commented:
目前的邏輯是所有參數必須同時都存在，才算符合規則:
```js
                    // 確認所有 keys 都存在
                    const allKeysExist = keys.every(key => parsedUrl.searchParams.has(key));

                    if (domainMatch && allKeysExist) {
                        keys.forEach(key => {
                            parsedUrl.searchParams.delete(key);
                        });
                    }
```

原本是有想說連同參數順序也要符合，不過後來並沒做到，也覺得似乎並非絕對需要。

該規則不加上域名的原因，是因為 [Substack](https://substack.com/) 可以自訂域名，我遇到好幾個，剛剛查一下看到的列在下面:
- https://unchartedterritories.tomaspueyo.com/
- https://www.latent.space/
- https://www.exponentialview.co/

想說 Substack 上自訂域名的網站應該數量也不少，若要求六個參數必須同時都存在，應該很難會有其他不相干網站會符合才對。若以後遇到一個就加一個發一個 PR，您跟其他使用者是否會覺得，不時就加一個很少人閱讀的網站，比對的參數也都是重覆的這六個，而且又會讓所有的瀏覽器 URL 多次呼叫 `removeByDomain()`，感到很麻煩又沒效率? 不知以上我的判斷您是否認同呢?

另我認為先前既有的一些規則，可能也可以用這個 `removeByDomainWithKeys()` 替代。但我自己沒有完整的網址去重覆測試，也不清楚是否有一些例外狀況，故不敢照己意去亂動。

doggy8088 commented:
六個同時都相同的網站應該不容易遇到沒錯，這點我同意。

你新增的這個 function 我覺得沒那麼直觀，所以乍看之下無法理解是「六個都有」才會移除，我覺得函式名稱與參數傳入的部分，應該都可以調整一下。例如：

```
.removeByDomainThatMatchAllKeys('*.substack.com', ['publication_id', 'post_id', 'isFreemail', 'r', 'token', 'triedRedirect'])
```

另外，為了自訂域名而套用到所有網站，我依然覺得不妥。

ChrisTorng committed:
TrackingTokenStripper.user.js 之 removeByDomainWithKeys() 改名 removeByDomainThatMatchAllKeys()，參數 keys 改收陣列。
原 Substack 相關移除項目不指定 domain，改指定 *.substack.com/unchartedterritories.tomaspueyo.com/www.latent.space 三項。

ChrisTorng commented:
已依建議修改:
> TrackingTokenStripper.user.js 之 `removeByDomainWithKeys()` 改名 `removeByDomainThatMatchAllKeys()`，參數 `keys` 改收陣列。
原 Substack 相關移除項目不指定 domain，改指定 *.substack.com/unchartedterritories.tomaspueyo.com/www.latent.space 三項。

我沒在看的網站就沒列上去。

不過我認為原有的 `remove()` 項目也不少，而且都套用在所有域名，比如:
```js
            .remove('mc') // sendgrid.com
            .remove('mcd') // sendgrid.com
            .remove('cvosrc') // sendgrid.com
```
感覺檢查條件比我加的這幾個更寬鬆? 是否也是為了自訂域名所以才不檢查? 不知是否三個必定同時出現?

另我再仔細看一下 `remove()` 實作，認為:
```js
                        if (value && value === parsedUrl.searchParams.get(name)) {
                            parsedUrl.searchParams.delete(name);
                        }
                        if (!value) {
                            parsedUrl.searchParams.delete(name);
                        }
```
可以合併為:
```js
                        if (!value || value === parsedUrl.searchParams.get(name)) {
                            parsedUrl.searchParams.delete(name);
                        }
```
